src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioMetadataAsSourceFileLSPService.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices.Utilities;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation
+{
+    internal class VisualStudioMetadataAsSourceFileLSPService : IMetadataAsSourceFileLSPService
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+
+        public VisualStudioMetadataAsSourceFileLSPService(SVsServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+
+            var componentModel = IServiceProviderExtensions.GetService<SComponentModel, IComponentModel>(_serviceProvider);
+            _metadataAsSourceFileService = componentModel.GetService<IMetadataAsSourceFileService>();
+        }
+
+        public async Task<MetadataAsSourceFile?> GetAndOpenGeneratedFileAsync(ISymbol symbol, Project project, CancellationToken cancellationToken)
+        {
+            var result = await _metadataAsSourceFileService.GetGeneratedFileAsync(
+                project, symbol, allowDecompilation: false, cancellationToken).ConfigureAwait(false);
```

Comment:
Up to you, as I'm not too fussed about it, but I was imagining that this call would stay where it was in the handler, and this service would just have an OpenFileAsync method, and just be `IVisualStudioOpenDocumentService` or some name. That was largely based on the assumption that _something_ at some point would need to also open documents. Up to you though, I don't feel too strongly about it.

Reply:
Ah, I think that would definitely make things cleaner. I've updated the code to reflect your suggestion. Thanks!